### PR TITLE
fix Issue 22133 - [REG2.097] Breaking change in DotTemplateExp type semantics leading to e.g. isInputRange regression

### PR DESCRIPTION
**Pull request by ibuclaw**
Reverts dlang/dmd#12294

This change had other surprising side effects.

```
import std.range.primitives : isInputRange;
struct Slice
{
    bool empty() const;
    int front() const;
    void popFront()()
    {
    }
}
static assert(isInputRange!(      Slice) == true);
static assert(isInputRange!(const Slice) == false);  // fails since 2.097.0
```

### CODE REVIEW

**Comment by ibuclaw** [comment]
Added regression test, linked to newly created issue 22133.

**Review by kinke — DISMISSED** [review]

**Comment by kinke** [comment]
Also pinging @MartinNowak - this is IMO crucial and a blocker for DMD v2.097.1 final.

**Comment by kinke** [comment]
At least 2 more PRs depend on the original PR: https://github.com/dlang/dmd/pull/12295 and https://github.com/dlang/dmd/pull/12544

**Comment by ibuclaw** [comment]
In this code:
```
static assert(isInputRange!(      Slice) == true);
```
This branch is hit, and resolveProperties returns a non-Error.
https://github.com/dlang/dmd/blob/ce1c1874c38d58d59586c3670ca95912746a692f/src/dmd/expressionsem.d#L1291-L1313
In this code:
```
static assert(isInputRange!(const Slice) == false);
```
This branch is hit, or at least was hit until the regressing PR was added.  Now an error never occurs because the `type` is set to `Type.tvoid`.
https://github.com/dlang/dmd/blob/ce1c1874c38d58d59586c3670ca95912746a692f/src/dmd/expressionsem.d#L1361-L1365

**Comment by ibuclaw** [comment]
Viewing the regression test at another angle - from that of the is/__traits constraint itself.  Ultimately what we have is:
```
struct Slice
{
    void popFront()() {}
}
auto test22133a()
{
    const Slice ret;
    ret.popFront;        // error: template has no effect.
    return ret.popFront; // not resolved, but no error
}
auto test22133b()
{
    Slice ret;
    ret.popFront;        // resolved as callexp
    return ret.popFront; // likewise.
}
```
So that suggests that the fix that doesn't demand a full revert is to call `discardValue` in the ReturnStatement semantic too.

**Comment by ibuclaw** [comment]
Waiting for what the testsuite says...

**Comment by ibuclaw** [comment]
https://github.com/dlang-community/libdparse/blob/fbdf71b929a9f9bc982d0204f921371806e5d412/src/dparse/stack_buffer.d#L76

So will have to split off the broader change to master with deprecation process.

**Comment by kinke** [comment]
So one of the `discardValue` tasks is to reject unresolved (=> newly void-typed) `DotTemplateExp` by emitting a cryptic error msg and pretending to have no side effects? Hmm, not very elegant IMO. ;)

Looking at one of the adapted testcases in the original PR, I cannot help but thinking this void-typing is seriously flawed, here a little adaptation passing with v2.097:
```D
int foo()() { return 0; }

struct B
{
    int foo()() { return 0; }
}

void bar()
{
    static assert(is(typeof(foo) == void));
    auto i1 = foo;
    static assert(is(typeof(i1) == int));

    static assert(is(typeof(B().foo) == void));
    auto i2 = B().foo;
    static assert(is(typeof(i2) == int));
}
```

**Comment by kinke** [comment]
Applying `discardValue()` to void-typed return statement expressions makes IMO sense though in general, preventing nonsensical oversights as in libdparse.

**Comment by BorisCarvajal** [comment]
> Looking at one of the adapted testcases in the original PR, I cannot help but thinking this void-typing is seriously flawed,

I agree, there is an overuse of `void` in the compiler internals, there are many objects whose type is `void` when it could be something else like an `unresolved` or `none` type, having this distinction should help in general.

This bug is similar to #12192, in which case the orphaned void-typed object is a `ScopeExp`, where again only `discardValue` is able to get rid of it with an error.
So in the current state, calling `discardValue` is an acceptable workaround, it's very unfortunate that some code is failing.

**Comment by ibuclaw** [comment]
I've just tried compiling `return ret.popFront;` down to object code in dmd, and I get:
```
[test22133.d(23)] dottd: ret.popFront()()
core.exception.AssertError@src/dmd/e2ir.d(1196): Assertion failure
----------------
??:? _d_assertp [0x55b6d267d3f1]
src/dmd/e2ir.d:1196 _ZN6toElem13ToElemVisitor5visitEP10Expression [0x55b6d254cdc8]

```
So yet another regression that #12294 introduced.

**Comment by ibuclaw** [comment]
> So one of the `discardValue` tasks is to reject unresolved (=> newly void-typed) `DotTemplateExp` by emitting a cryptic error msg and pretending to have no side effects? Hmm, not very elegant IMO. ;)

Before #12294 the error was: `Error: cannot resolve type for ret.popFront()()`.  It is just lucky that exp statements catch unresolved templates.

**Comment by BorisCarvajal** [comment]
`TemplateExp` and `ScopeExp` has these:
```d
    override bool checkType()
    {
        error("%s `%s` has no type", td.kind(), toChars());
        return true;
    }

    override bool checkValue()
    {
        error("%s `%s` has no value", td.kind(), toChars());
        return true;
    }

```
I think adding them to `DotTemplateExp` should work at least in the `return` case.

**Comment by ibuclaw** [comment]
> `TemplateExp` and `ScopeExp` has these:

Incidentally, I'm about 10 minutes ahead of you and have already done this locally. ;-)

**Comment by ibuclaw** [comment]
Moved deprecation of return value with no side effects to #12898.

**Review by kinke — APPROVED** [review]
Nice, much better now IMO.

File: test/fail_compilation/fail22133.d [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// https://issues.dlang.org/show_bug.cgi?id=22133
+/*
+TEST_OUTPUT
+---
+fail_compilation/fail22133.d(16): Error: `s.popFront()()` has no effect
```
Comment (kinke) [inline test/fail_compilation/fail22133.d:5]: As a follow-up, we could add a `TOK.dotTemplateDeclaration` case in `discardValue()` and offer a better error message.

**Review by BorisCarvajal — APPROVED** [review]

**Comment by MartinNowak** [comment]
If this is bad enough, we can do another point release @kinke. Just ping me, otherwise would be included in the next release in 2-3 weeks.

**Comment by kinke** [comment]
@MartinNowak: It's blocking the compiler bump to v2.097 for Symmetry, as `mir-algorithm` uses such a construct: https://github.com/libmir/mir-algorithm/blob/d59f9c1de71d436ff6a7787744922d97a1a61f56/source/mir/ndslice/slice.d#L1625. - Regression testing takes time (especially this time with v2.097), as e.g. several dub packages might need to be adapted as a prerequisite, so that we usually go with the first point release - meaning that the bump would be delayed by ~2 months for v2.098.1.

TLDR: An official v2.097.2 including this fix would be awesome for us. But I'll first check whether this fix is really the last one we need. ;)

**Comment by ibuclaw** [comment]
PR title updated.  Let me just check the test that @kinke most recently posted.

**Comment by ibuclaw** [comment]
> So one of the `discardValue` tasks is to reject unresolved (=> newly void-typed) `DotTemplateExp` by emitting a cryptic error msg and pretending to have no side effects? Hmm, not very elegant IMO. ;)
> 
> Looking at one of the adapted testcases in the original PR, I cannot help but thinking this void-typing is seriously flawed, here a little adaptation passing with v2.097:
> 
> ```d
> int foo()() { return 0; }
> 
> struct B
> {
>     int foo()() { return 0; }
> }
> 
> void bar()
> {
>     static assert(is(typeof(foo) == void)); // resolution failure
>     auto i1 = foo; // resolution succeeds
>     static assert(is(typeof(i1) == int));
> 
>     static assert(is(typeof(B().foo) == void));
>     auto i2 = B().foo;
>     static assert(is(typeof(i2) == int));
> }
> ```

@kinke should I check this in just to document in the testsuite what is the current behaviour?  Possibly with a note that this looks questionable and suggestion of what the correct behaviour should be if someone changes it.

**Comment by kinke** [comment]
Feel free to - FWIW, it's an adapation of `test/compilable/test21742.d`.

**Comment by kinke** [comment]
@MartinNowak: This was apparently the last blocker, so green light for v2.097.2 from my side. :)
